File: android/app/src/main/java/com/kvkleh/sbtsurvey/ui/components/StatusComponents.kt
package com.kvkleh.sbtsurvey.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CloudOff
import androidx.compose.material.icons.filled.ErrorOutline
import androidx.compose.material.icons.filled.GpsFixed
import androidx.compose.material.icons.filled.GpsNotFixed
import androidx.compose.material.icons.filled.GpsOff
import androidx.compose.material.icons.filled.Save
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import com.kvkleh.sbtsurvey.data.location.GpsStatus

/** What the auto-save pill is currently reporting. */
enum class AutoSaveState { Idle, Saving, Saved, Failed }

/**
 * The small pill in the app bar that tells the surveyor their work is on disk.
 * Auto-save is the app's central promise, so it is always visible while a form
 * is open rather than hidden behind a toast.
 */
@Composable
fun AutoSaveIndicator(state: AutoSaveState, modifier: Modifier = Modifier) {
    val alpha by animateFloatAsState(
        targetValue = if (state == AutoSaveState.Idle) 0.75f else 1f,
        label = "autoSaveAlpha"
    )
    val (label, icon, tint) = when (state) {
        AutoSaveState.Idle -> Triple("Auto-save on", Icons.Filled.Save, MaterialTheme.colorScheme.onPrimary)
        AutoSaveState.Saving -> Triple("Saving…", null, MaterialTheme.colorScheme.onPrimary)
        AutoSaveState.Saved -> Triple("Auto-saved", Icons.Filled.Save, MaterialTheme.colorScheme.onPrimary)
        AutoSaveState.Failed -> Triple("Not saved", Icons.Filled.ErrorOutline, MaterialTheme.colorScheme.onPrimary)
    }

    StatusPill(
        label = label,
        icon = icon,
        contentColor = tint,
        containerColor = if (state == AutoSaveState.Failed) {
            MaterialTheme.colorScheme.error
        } else {
            Color.White.copy(alpha = 0.18f)
        },
        showSpinner = state == AutoSaveState.Saving,
        modifier = modifier.alpha(alpha)
    )
}

/** Status of the GPS receiver, shown on the location card. */
@Composable
fun GpsStatusPill(status: GpsStatus, modifier: Modifier = Modifier) {
    val (label, icon, container, content) = when (status) {
        is GpsStatus.Ready -> StatusStyle(
            "GPS Ready",
            Icons.Filled.GpsFixed,
            MaterialTheme.colorScheme.secondaryContainer,
            MaterialTheme.colorScheme.onSecondaryContainer
        )

        GpsStatus.Acquiring -> StatusStyle(
            "Acquiring GPS…",
            Icons.Filled.GpsNotFixed,
            MaterialTheme.colorScheme.tertiaryContainer,
            MaterialTheme.colorScheme.onTertiaryContainer
        )

        GpsStatus.PermissionRequired -> StatusStyle(
            "Location permission needed",
            Icons.Filled.GpsOff,
            MaterialTheme.colorScheme.errorContainer,
            MaterialTheme.colorScheme.onErrorContainer
        )

        GpsStatus.ServicesDisabled -> StatusStyle(
            "Location services off",
            Icons.Filled.GpsOff,
            MaterialTheme.colorScheme.errorContainer,
            MaterialTheme.colorScheme.onErrorContainer
        )

        GpsStatus.Unavailable -> StatusStyle(
            "Location unavailable",
            Icons.Filled.CloudOff,
            MaterialTheme.colorScheme.errorContainer,
            MaterialTheme.colorScheme.onErrorContainer
        )
    }

    StatusPill(
        label = label,
        icon = icon,
        contentColor = content,
        containerColor = container,
        showSpinner = status == GpsStatus.Acquiring,
        modifier = modifier
    )
}

private data class StatusStyle(
    val label: String,
    val icon: ImageVector,
    val container: Color,
    val content: Color
)

@Composable
fun StatusPill(
    label: String,
    icon: ImageVector?,
    contentColor: Color,
    containerColor: Color,
    modifier: Modifier = Modifier,
    showSpinner: Boolean = false
) {
    Row(
        modifier = modifier
            .background(containerColor, RoundedCornerShape(50))
            .padding(horizontal = 10.dp, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center
    ) {
        if (showSpinner) {
            CircularProgressIndicator(
                modifier = Modifier.size(14.dp),
                strokeWidth = 2.dp,
                color = contentColor
            )
            Spacer(Modifier.width(6.dp))
        } else if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = contentColor,
                modifier = Modifier.size(16.dp)
            )
            Spacer(Modifier.width(6.dp))
        }
        Text(
            text = label,
            style = MaterialTheme.typography.labelMedium,
            color = contentColor
        )
    }
}

File: ios/SeabuckthornSurvey/UI/Components/StatusComponents.swift
import SwiftUI

/// What the auto-save pill is currently reporting.
enum AutoSaveState {
    case idle
    case saving
    case saved
    case failed
}

/// The pill in the navigation bar that tells the surveyor their work is on disk.
/// Auto-save is the app's central promise, so it stays visible while a form is open
/// rather than hiding behind a toast.
struct AutoSaveIndicator: View {

    let state: AutoSaveState

    var body: some View {
        HStack(spacing: 6) {
            switch state {
            case .saving:
                ProgressView().controlSize(.mini).tint(SBTColor.inkSoft)
            case .failed:
                Image(systemName: "exclamationmark.triangle.fill")
            default:
                Image(systemName: "checkmark.circle.fill")
            }
            Text(label)
                .font(.caption.weight(.medium))
        }
        .foregroundStyle(state == .failed ? Color.white : SBTColor.inkSoft)
        .padding(.horizontal, 10)
        .padding(.vertical, 5)
        .background(
            state == .failed ? SBTColor.danger : SBTColor.surfaceDim,
            in: Capsule()
        )
        .accessibilityLabel(label)
    }

    private var label: String {
        switch state {
        case .idle: return "Auto-save on"
        case .saving: return "Saving…"
        case .saved: return "Auto-saved"
        case .failed: return "Not saved"
        }
    }
}

/// Status of the GPS receiver, shown on the location card.
struct GpsStatusPill: View {

    let status: GpsStatus

    var body: some View {
        HStack(spacing: 6) {
            if status == .acquiring {
                ProgressView().controlSize(.mini).tint(SBTColor.onTertiaryContainer)
            } else {
                Image(systemName: symbol)
            }
            Text(label).font(.caption.weight(.medium))
        }
        .foregroundStyle(foreground)
        .padding(.horizontal, 10)
        .padding(.vertical, 5)
        .background(background, in: Capsule())
        .accessibilityLabel(label)
    }

    private var label: String {
        switch status {
        case .ready: return "GPS Ready"
        case .acquiring: return "Acquiring GPS…"
        case .permissionRequired: return "Location permission needed"
        case .servicesDisabled: return "Location services off"
        case .unavailable: return "Location unavailable"
        }
    }

    private var symbol: String {
        switch status {
        case .ready: return "location.fill"
        case .acquiring: return "location"
        case .permissionRequired, .servicesDisabled: return "location.slash"
        case .unavailable: return "exclamationmark.icloud"
        }
    }

    private var foreground: Color {
        switch status {
        case .ready: return SBTColor.onSecondaryContainer
        case .acquiring: return SBTColor.onTertiaryContainer
        default: return Color.white
        }
    }

    private var background: Color {
        switch status {
        case .ready: return SBTColor.secondaryContainer
        case .acquiring: return SBTColor.tertiaryContainer
        default: return SBTColor.danger
        }
    }
}

/// The block of optional fields a record is missing, shown before saving.
struct MissingDataNotice: View {

    let warnings: [String]

    var body: some View {
        if warnings.isEmpty {
            EmptyView()
        } else {
            VStack(alignment: .leading, spacing: 6) {
                Text("Optional information not recorded")
                    .font(.subheadline.weight(.semibold))
                ForEach(warnings, id: \.self) { warning in
                    Text("• \(warning)").font(.subheadline)
                }
                Text("The survey can still be saved. These fields can be filled in later by editing the record.")
                    .font(.caption)
                    .padding(.top, 4)
            }
            .foregroundStyle(SBTColor.onTertiaryContainer)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(16)
            .background(
                SBTColor.tertiaryContainer,
                in: RoundedRectangle(cornerRadius: SBTMetrics.cornerRadius)
            )
        }
    }
}
